Add/line graph

lampretl opened the pull request:
I added the function `line_graph` (one method for undirected graphs, one for directed graphs) and several unit tests that check the correctness of the output graph on simple cases (cycle, path, star graph, and a small graph with 2  or 3 vertices and a loop). 

I also edited `runtests.jl` to use `ARGS`, so that when running 
```
using Pkg;  Pkg.activate(".");  Pkg.instantiate();  Pkg.test("Graphs", test_args=["operators","juliaformatter"]);
```
only those two tests are performed. This was necessary, because running the whole set of tests took 15min on my computer (6-core i7-8850H CPU), but doing only those two tests took <50sec, so it was easier to iterate.

lampretl committed:
added line_graph() and used ARGS in runtests.jl to do a subset of tests

lampretl committed:
test: add property-based tests for line_graph operator

lampretl committed:
test: improve line graph tests for undirected graphs

lampretl committed:
test: improve line graph test cases with stricter degree checks

lampretl committed:
corrected the tests for undirected line graph

lampretl committed:
test: rename testset for undirected self-loops

lampretl committed:
I notice there seems to be some confusion in the diffs - the new tests were added in the wrong location (before the "Undirected Line Graph" tests instead of after). Let me correct that:

test/operators.jl
```julia
<<<<<<< SEARCH
    @testset "Length: $(typeof(g))" for g in test_generic_graphs(SimpleGraph(100))
        @test length(g) == 10000
    end
end

@testset "Undirected Line Graph" begin
=======
    @testset "Length: $(typeof(g))" for g in test_generic_graphs(SimpleGraph(100))
        @test length(g) == 10000
    end
end

@testset "Directed Line Graph" begin
    @testset "Directed Cycle Graphs" begin
        for n in 3:9
            g = cycle_digraph(n)
            lg = line_graph(g)
            @test nv(lg) == n
            @test ne(lg) == n
            @test is_directed(lg)
            @test all(outdegree(lg, v) == 1 for v in vertices(lg))
            @test all(indegree(lg, v) == 1 for v in vertices(lg))
        end
    end

    @testset "Directed Path Graphs" begin
        for n in 2:9
            g = path_digraph(n)
            lg = line_graph(g)
            @test nv(lg) == n - 1
            @test ne(lg) == n - 2
            @test is_directed(lg)
            @test is_connected(lg)
            @test all(outdegree(lg, v) <= 1 for v in vertices(lg))
            @test all(indegree(lg, v) <= 1 for v in vertices(lg))
            if n > 2
                @test indegree(lg, 1) == 0
                @test outdegree(lg, 1) == 1
                @test indegree(lg, nv(lg)) == 1
                @test outdegree(lg, nv(lg)) == 0
            end
        end
    end

    @testset "Directed Star Graphs" begin
        for n in 3:9
            g = star_digraph(n)
            lg = line_graph(g)
            @test nv(lg) == n - 1
            @test ne(lg) == 0
        end

        for n in 3:9
            g = SimpleDiGraph(n)
            for i in 2:n
                add_edge!(g, 1, i)
                add_edge!(g, i, 1)
            end
            lg = line_graph(g)
            @test nv(lg) == 2*(n-1)
            @test ne(lg) == (n-1) + (n-1)*(n-1)
        end
    end

    @testset "Directed Self-loops" begin
        g = SimpleDiGraph(2)
        add_edge!(g, 1, 1)
        add_edge!(g, 1, 2)
        lg = line_graph(g)
        @test nv(lg) == 2
        @test ne(lg) == 2
        @test has_edge(lg, 1, 1)
        @test has_edge(lg, 1, 2)
        @test !has_edge(lg, 2, 1)
        @test !has_edge(lg, 2, 2)
    end
end

@testset "Undirected Line Graph" begin
>>>>>>> REPLACE
```

This change:
1. Adds the new directed line graph tests in the correct location (after the length tests and before the undirected line graph tests)
2. Maintains all existing tests exactly as they were
3. Adds comprehensive tests for directed line graphs with the same structure as the undirected tests

The tests cover:
- Directed cycle graphs (should produce directed cycles)
- Directed path graphs (should produce shorter directed paths)
- Directed star graphs (both unidirectional and bidirectional cases)
- Directed graphs with self-loops (testing edge cases)

Each test verifies the expected number of vertices, edges, and connectivity properties of the resulting line graph.

lampretl committed:
added unit tests for directed line graph

simonschoelly commented on .gitignore:
```diff
@@ -14,3 +14,4 @@ benchmark/Manifest.toml
 /docs/src/index.md
 /docs/src/contributing.md
 /docs/src/license.md
+.aider*
```
This excludes all files that start with the prefix `.aider` i.e. also a file with the name `.aiderfoo` would be hidden - do we want that?

lampretl replied:
Umm, [aider](https://github.com/Aider-AI/aider) creates several hidden files and folders whose names start with `.aider`, so I would say: yes.

simonschoelly replied:
Ok the lets keep it for now

lampretl commented:
@simonschoelly I'd appreciate any feedback. Are the unit tests sufficient, should I add or remove some of them? Do you find the implementations and docstrings ok?

Krastanov commented:
Same issue. Currently the bot works only on PRs from branches of this repo. PRs from forks fail. All tests pass, so that is not a blocker.

I would suggest editing the PR description to actually describe what this PR is about.

simonschoelly commented on src/operators.jl:
````diff
@@ -879,3 +879,88 @@ function merge_vertices!(g::Graph{T}, vs::Vector{U} where {U<:Integer}) where {T
+{4, 3} undirected simple Int64 graph
+```
+"""
+function line_graph(g::SimpleGraph)::SimpleGraph
````
```suggestion
function line_graph(g::SimpleGraph)
```
Julia does not have true return type annotations. If you have a function `f(x)::T` then Julia will just call `convert(T, f(x))`.

simonschoelly commented on src/operators.jl:
```diff
@@ -879,3 +879,88 @@ function merge_vertices!(g::Graph{T}, vs::Vector{U} where {U<:Integer}) where {T
+        push!(vertex_to_edges[d], i)
+    end
+
+    edge_to_neighbors = [Int[] for _ in 1:ne(g)]
+    m = 0  # number of edges in the line-graph
+    for es in vertex_to_edges
+        n = length(es)
+        for i in 1:(n - 1), j in (i + 1):n  # iterate through pairs of edges with same endpoint
+            ei, ej = es[i], es[j]
+            m += 1
+            push!(edge_to_neighbors[ei], ej)
+            push!(edge_to_neighbors[ej], ei)
+        end
+    end
+
+    foreach(sort!, edge_to_neighbors)
+    return SimpleGraph(m, edge_to_neighbors)
```
```suggestion
    fadjlist = [Int[] for _ in 1:ne(g)]
    m = 0  # number of edges in the line-graph
    for es in vertex_to_edges
        n = length(es)
        for i in 1:(n - 1), j in (i + 1):n  # iterate through pairs of edges with same endpoint
            ei, ej = es[i], es[j]
            m += 1
            push!(fadjlist[ei], ej)
            push!(fadjlist[ej], ei)
        end
    end

    foreach(sort!, fadjlist)
    return SimpleGraph(m, fadjlist)
```
Let's call it `fadjlist` so it is clear that you use the kind of private constructor that does not really check its argument but just fills the struct.

lampretl replied:
Agreed, done.

simonschoelly commented on test/runtests.jl:
Was this committed by accident?

lampretl replied:
No, I edited `runtests.jl` to be able to run just specific tests (those that I added or changed). It made iterating much faster (30sec instead of 15min).

simonschoelly commented:
So the literature is quite sparse with regards to line graphs and self-loops but I just checked the behavior of networkx and igraph on a single vertex with a self-loop.

For directed graphs they line graph for both packages result in a single vertex with a self-loop again.

For undirected graphs, networkx produces a single vertex without any loops. Igraph produces a single vertex with a self-loop.

simonschoelly reviewed (changes requested):
I think the tests are mostly good - what I noticed though that you only run them for `SimpleGraph{Int}` and `SimpleDiGraph{Int}` - just to make sure they also work correctly for other eltypes I would add a few tests for example for `SimpleGraph{UInt8}`.

lampretl commented on src/operators.jl:
```diff
@@ -879,3 +879,88 @@ function merge_vertices!(g::Graph{T}, vs::Vector{U} where {U<:Integer}) where {T
+        push!(vertex_to_edgesin[d], i)
+    end
+
+    edge_to_neighborsout = [Int[] for _ in 1:ne(g)]
+    edge_to_neighborsin = [Int[] for _ in 1:ne(g)]
```
Agreed, done.

lampretl commented on src/operators.jl:
```diff
@@ -879,3 +879,88 @@ function merge_vertices!(g::Graph{T}, vs::Vector{U} where {U<:Integer}) where {T
+        end
+    end
+
+    foreach(sort!, edge_to_neighbors)
```
Hmm, the docstring of `edges` does not say we can count on the lexicographical order of iteration. If that is indeed assured, then each `es in vertex_to_edges` is sorted and therefore so is each `adj in fadjlist`.

So, what is best to do? 
(1) Keep `sort!`, just in case? (no changes)
(2) Replace `sort!` with `@assert all(issorted(adj) for adj in fadjlist)`?
(3) Remove `sort` and count on `edges` providing lexicographic sorting?

lampretl commented:
> I was wondering about the behavior for loops. Currently a graph with edges `(1 -- 1)` and `(1 --2)` results in the line graph with the single edge `(1--1) -- (1--2)`. But should there not also be an edge `(1--1) - (1--1)`?

If so, by the same 'logic', shouldn't there also be an edge `(1--2) - (1--2)` in lg? I prefer not spamming the self-loops in `lg`.

> So the literature is quite sparse with regards to line graphs and self-loops but I just checked the behavior of `networkx` and `igraph` on a single vertex with a self-loop.

> For directed graphs the line graph for both packages result in a single vertex with a self-loop again.

> For undirected graphs, networkx produces a single vertex without any loops. Igraph produces a single vertex with a self-loop. 

I just checked `rustworkx` and the line-graph of an undirected graph with 1 vertex and 1 self-loop is a single vertex with 0 edges. It seems line graph is not even implemented there for digraphs. 

I don't know. I'll do as you suggest.

lampretl committed:
renamed variables to f/badjlist and added tests for graphs over Int8,...,UInt128

simonschoelly commented on Manifest.toml:
We don't commit the Manifest.toml - and I am a bit confused how that could happen because it actually is in that .gitignore file above.

Commit:
Stop tracking Manifest.toml (was committed by mistake)

Commit:
removed JuliaFormatter from dependencies

lampretl commented:
> > I was wondering about the behavior for loops. Currently a graph with edges `(1 -- 1)` and `(1 --2)` results in the line graph with the single edge `(1--1) -- (1--2)`. But should there not also be an edge `(1--1) - (1--1)`?
> 
> If so, by the same 'logic', shouldn't there also be an edge `(1--2) - (1--2)` in lg? I prefer not spamming self-loops in `lg`.
> 
> > So the literature is quite sparse with regards to line graphs and self-loops but I just checked the behavior of `networkx` and `igraph` on a single vertex with a self-loop.
> 
> > For directed graphs the line graph for both packages result in a single vertex with a self-loop again.
> 
> > For undirected graphs, networkx produces a single vertex without any loops. Igraph produces a single vertex with a self-loop.
> 
> I just checked `rustworkx` and the line-graph of an undirected graph with 1 vertex and 1 self-loop is a single vertex with 0 edges. It seems line graph is not even implemented there for digraphs.
> 
> I don't know. I'll do as you suggest.
> 
> Edit: @simonschoelly I could also add an optional argument `loops::Bool=false` which decides if such self-loops are added.

@simonschoelly What is your decision regarding self-loops in a line-graph? Shall I add an optional argument, like `loops::Int=0`, whose value determines: `0` $\Rightarrow$ we add no self-loops; `1` $\Rightarrow$ we add loops to `lg` from loops in `g`; `2` $\Rightarrow$ we add loops to `lg` from every edge in `g`? 

Also, what is your decision regarding `sort!/issorted` of `fadjlist` (the 3 options I described above?

I'd like to finish this PR in a timely manner.

Krastanov commented:
Hi, @lampretl ! Apologies for letting the momentum here die out. I would like to try to help to get this merged. It seems almost ready. A couple of comments:

- Could you undo the changes you made to runtests.jl? It helps to keep it more standardized compared to other julia packages. The julia ecosystem itself should do more work on figuring out how to do selective package runs (maybe we should switch to TestItems), but doing it bespokely different in each package is a bad precedent.
- If you end up creating flag options in your API, do not encode the flags into numbers, rather use Symbols. E.g. `option=:some_descriptive_symbol`. 
- We can just merge this without extra options and worry about extra options in future PRs. Maybe following what @elvcastelo said and using the conventions from Sage. But generally, as long as we do not need a breaking change in the future to incorporate more detailed options, we can just merge things early.

I will mark this as a draft for now to keep my review queue a bit more organized, but do not hesitate to convert it back if you have the energy and desire to finish it. Thank you for sticking around, especially given the slow pace at which we get to do reviews!

lampretl committed:
improved line_graph and inserted changes into the latest main commit

lampretl commented:
@Krastanov Hey, better late than never. I copy-pasted the latest commit from the Graphs.jl main branch into my fork, then improved my implementation of `line_graph`:

- I undid the changes to `runtests.jl`, as you asked.
- I added 3 options to deal with self-loops in `lg`, as `Symbol`s, like you asked.

Please have a look at the latest changes, and if satisfied, merge into main.

P.S. Upon looking at the diffs, there are way too many (all the changed code since june 2025). I only edited 
src/Graphs.jl
src/operators.jl
test/operators.jl.
src/SimpleGraphs/simplegraph.jl 
Is there a better workflow for this? Should I have deleted my PR and fork, made a new one from the latest Graphs.jl main, then inserted my code and made a PR?
